### PR TITLE
[SC-362] April 3, 2024 Spell

### PR DESCRIPTION
**Pull request by hexonaut**
Previous spell still needs to pass and going to add more to the kill switch integration test, but the rest is ready for review.

### CODE REVIEW

File: src/proposals/20240403/SparkEthereum_20240403.t.sol [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        loadPoolContext(poolAddressesProviderRegistry.getAddressesProvidersList()[0]);
+    }
+
+    function testKillSwitchActivation() public {
+        IKillSwitchOracle kso = IKillSwitchOracle(KILL_SWITCH_ORACLE);
+
+        assertEq(aclManager.isRiskAdmin(KILL_SWITCH_ORACLE), false);
+        assertEq(kso.numOracles(), 0);
+        assertEq(kso.oracleThresholds(WBTC_BTC_ORACLE),  0);
+        assertEq(kso.oracleThresholds(STETH_ETH_ORACLE), 0);
```
Comment (lucas-manuel) [inline src/proposals/20240403/SparkEthereum_20240403.t.sol:50]: Align

File: src/proposals/20240403/SparkEthereum_20240403.t.sol [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+        GovHelpers.executePayload(vm, payload, executor);
+
+        assertEq(aclManager.isRiskAdmin(KILL_SWITCH_ORACLE), true);
+        assertEq(kso.numOracles(), 2);
+        assertEq(kso.oracleThresholds(WBTC_BTC_ORACLE),  0.95e8);
+        assertEq(kso.oracleThresholds(STETH_ETH_ORACLE), 0.95e18);
```
Comment (lucas-manuel) [inline src/proposals/20240403/SparkEthereum_20240403.t.sol:57]: Align

File: src/proposals/20240403/SparkEthereum_20240403.t.sol [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+        // Replace the aggregator with a mock
+        MockAggregator wbtcAggregator = new MockAggregator(0.95e8);
+        vm.store(
+            WBTC_BTC_ORACLE,
+            bytes32(uint256(2)),
+            bytes32((uint256(uint160(address(wbtcAggregator))) << 16) | 1)
+        );
+
+        assertEq(IChainlinkAggregator(WBTC_BTC_ORACLE).latestAnswer(),  0.95e8);
```
Comment (lucas-manuel) [inline src/proposals/20240403/SparkEthereum_20240403.t.sol:80]: Extra space

File: src/proposals/20240403/SparkEthereum_20240403.t.sol [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        assertEq(_getBorrowEnabled(SDAI),   false);
+        assertEq(_getBorrowEnabled(USDC),   true);
+        assertEq(_getBorrowEnabled(WETH),   true);
+        assertEq(_getBorrowEnabled(WSTETH), true);
+        assertEq(_getBorrowEnabled(WBTC),   true);
+        assertEq(_getBorrowEnabled(GNO),    false);
+        assertEq(_getBorrowEnabled(RETH),   true);
+        assertEq(_getBorrowEnabled(USDT),   true);
+
+        kso.trigger(WBTC_BTC_ORACLE);
```
Comment (lucas-manuel) [inline src/proposals/20240403/SparkEthereum_20240403.t.sol:92]: Revert state and do the same for eth oracle

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Some comments but LGTM otherwise.

Checked:
- Links
- Diff
- Addresses in spell
- Bytecode of kill switch against v1.0.0 release
- Decimal values in oracles
- KillSwitchOracle interface

File: src/proposals/20240403/SparkEthereum_20240403.t.sol (hunk 1 above)
Comment (hexonaut) [reply src/proposals/20240403/SparkEthereum_20240403.t.sol:50]: fixed

File: src/proposals/20240403/SparkEthereum_20240403.t.sol (hunk 2 above)
Comment (hexonaut) [reply src/proposals/20240403/SparkEthereum_20240403.t.sol:57]: fixed

File: src/proposals/20240403/SparkEthereum_20240403.t.sol (hunk 3 above)
Comment (hexonaut) [reply src/proposals/20240403/SparkEthereum_20240403.t.sol:80]: fixed

File: src/proposals/20240403/SparkEthereum_20240403.t.sol (hunk 4 above)
Comment (hexonaut) [reply src/proposals/20240403/SparkEthereum_20240403.t.sol:92]: done

**Review by lucas-manuel — DISMISSED** [review]

**Review by lucas-manuel — APPROVED** [review]

**Comment by hexonaut** [comment]
Deployed.

**Review by barrutko — APPROVED** [review]
One minor comment, other that LGTM

File: src/proposals/20240403/SparkEthereum_20240403.sol [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        // Adjust Morpho Vault supply caps
+        IMetaMorpho(MORPHO_VAULT).submitCap(MarketParams({
+            loanToken:       DAI,
+            collateralToken: SUSDE,
+            oracle:          SUSDE_ORACLE,
+            irm:             MORPHO_DEFAULT_IRM,
+            lltv:            0.86e18
+        }), 200_000_000 ether);
+        IMetaMorpho(MORPHO_VAULT).submitCap(MarketParams({
```
Comment (lucas-manuel) [inline src/proposals/20240403/SparkEthereum_20240403.sol:65]: Apply to all

File: src/proposals/20240403/SparkEthereum_20240403.t.sol [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+    ) internal {
+        bytes32 id = _id(_config);
+        assertEq(IMetaMorpho(MORPHO_VAULT).config(id).cap, _currentCap);
+        PendingUint192 memory pendingCap = IMetaMorpho(MORPHO_VAULT).pendingCap(id);
+        if (_hasPending) {
+            assertEq(pendingCap.value, _pendingCap);
+            assertGt(pendingCap.validAt, 0);
+        } else {
+            assertEq(pendingCap.value, 0);
+            assertEq(pendingCap.validAt, 0);
```
Comment (lucas-manuel) [inline src/proposals/20240403/SparkEthereum_20240403.t.sol:294]: Align these

**Review by lucas-manuel — DISMISSED** [review]
Checked new addresses, checked new supply caps against forum post, reviewed tests, LGTM (left a formatting comment but not blocking merge on it)

File: src/interfaces/IMetaMorpho.sol [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    uint64  validAt;
+}
+
+interface IMetaMorpho {
+    function config(bytes32) external view returns (MarketConfig memory);
+    function submitCap(MarketParams memory marketParams, uint256 newSupplyCap) external;
+    function acceptCap(MarketParams memory marketParams) external;
+    function pendingCap(bytes32) external view returns (PendingUint192 memory);
+    function timelock() external view returns (uint256);
+}
```
[the remark this answers is not in the document]
Comment (hexonaut) [reply src/interfaces/IMetaMorpho.sol:29]: fixed

File: src/proposals/20240403/SparkEthereum_20240403.t.sol (hunk 6 above)
Comment (hexonaut) [reply src/proposals/20240403/SparkEthereum_20240403.t.sol]: done

**Review by barrutko — APPROVED** [review]

**Pull request merged**